snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/version/SnomedVersioningRequest.java

```diff
@@ -145,60 +146,75 @@ protected void doVersionComponents(TransactionContext context) throws Exception
+					componentIdsByReferringModule.put(description.getModuleId(), description.getConceptId());
+					componentIdsByReferringModule.put(description.getModuleId(), description.getTypeId());
+					componentIdsByReferringModule.put(description.getModuleId(), description.getCaseSignificanceId());
+					updatedComponent = SnomedDescriptionIndexEntry.builder(description);
+				} else if (componentToVersion instanceof SnomedRelationshipIndexEntry) {
+					final SnomedRelationshipIndexEntry relationship = (SnomedRelationshipIndexEntry) componentToVersion;
+					componentIdsByReferringModule.put(relationship.getModuleId(), relationship.getSourceId());
+					componentIdsByReferringModule.put(relationship.getModuleId(), relationship.getTypeId());
+					if (!relationship.hasValue()) {
+						// Values do not contribute to module dependencies
+						componentIdsByReferringModule.put(relationship.getModuleId(), relationship.getDestinationId());
+					}
+					componentIdsByReferringModule.put(relationship.getModuleId(), relationship.getModifierId());
+					componentIdsByReferringModule.put(relationship.getModuleId(), relationship.getCharacteristicTypeId());
+					updatedComponent = SnomedRelationshipIndexEntry.builder(relationship);
+				} else if (componentToVersion instanceof SnomedRefSetMemberIndexEntry) {
+					final SnomedRefSetMemberIndexEntry member = (SnomedRefSetMemberIndexEntry) componentToVersion;
+					componentIdsByReferringModule.put(member.getModuleId(), member.getReferenceSetId());
+					
+					registerIfConcept(componentIdsByReferringModule, member.getModuleId(), member.getReferencedComponentId());
+					
+					final Map<String, Object> additionalFields = member.getAdditionalFields();
+					SnomedRf2Headers.MEMBER_FIELDS_WITH_COMPONENT_ID.forEach(field -> {
+						registerIfConcept(componentIdsByReferringModule, member.getModuleId(), (String) additionalFields.get(field));
+					});
+					
+					updatedComponent = SnomedRefSetMemberIndexEntry.builder(member);
+				} else {
+					throw new UnsupportedOperationException("Not implemented case for: " + componentToVersion);
+				}
 				
-				updatedComponent = SnomedRefSetMemberIndexEntry.builder(member);
-			} else {
-				throw new UnsupportedOperationException("Not implemented case for: " + componentToVersion);
+				ctx.update(
+					componentToVersion, 
+					updatedComponent
+						.effectiveTime(effectiveTime)
+						.released(true)
+					.build()
+				);
 			}
-			
-			context.update(
-				componentToVersion, 
-				updatedComponent
-					.effectiveTime(effectiveTime)
-					.released(true)
-				.build()
-			);
-		}
 		
+			ctx.commit();
+		
+			if (!CompareUtils.isEmpty(componentIdsToPublish)) {
+				SnomedRequests.identifiers().preparePublish()
+					.setComponentIds(componentIdsToPublish)
+					.build()
+					.execute(ctx);
+			}
+		} catch (Exception e) {
+			throw new SnowowlRuntimeException(e);
```

Comment:
`ctx.rollback` should be called if an exception occurs. `TransactionalRequest` would handle this, but not for the manually managed "inner" transaction context instance.

Reply:
Side note: this could be done in `VersioningRequest` as well, as the capped transaction context is also locally managed, but it gets a bit more complicated in the general case as it might have already committed some pieces successfully, in which case rollback will be of no use. But what can be applied here is the error handling pattern from `TransactionalRequest`: 

```java
TransactionContext tx = null;
try (final TransactionContext transaction = context.openTransaction(context, author, "Versioning components", parentLockContext)) {
  tx = transaction;
  // Versioning work happens here
} catch (ApiException e) {
  if (tx != null) { tx.rollback(); }
  throw e;
} catch (Exception e) {
  if (tx != null) { tx.rollback(); }
  throw SnowowlRuntimeException.wrap(e);
}
```